presto-spi/src/main/java/com/facebook/presto/spi/Page.java

```diff
@@ -290,11 +293,16 @@ public Page prependColumn(Block column)
 
     private void updateRetainedSize()
     {
-        long retainedSizeInBytes = INSTANCE_SIZE + sizeOf(blocks);
+        AtomicLong retainedSizeInBytes = new AtomicLong(INSTANCE_SIZE + sizeOf(blocks));
+        Set<Object> referenceSet = Collections.newSetFromMap(new IdentityHashMap<>());
         for (Block block : blocks) {
-            retainedSizeInBytes += block.getRetainedSizeInBytes();
+            block.retainedBytesForEachPart((object, size) -> {
```

Comment:
My memory is that this is unbelievably expensive, which is why we only did this very sparingly.  Did you do some performance analysis?